CARDS-1344: As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires

marta- opened the pull request:
### Included in this PR:
* [CARDS-1344](https://phenotips.atlassian.net/browse/CARDS-1344): _As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires_ - 5/6 open sub-tasks
* [CARDS-1358](https://phenotips.atlassian.net/browse/CARDS-1358): _Create a mock patient authentication component_. - a patient identification screen and appropriate patient/visit subject creation. No actual authentication of a patient-type user.
* [CARDS-1349](https://phenotips.atlassian.net/browse/CARDS-1349): _Sync patient-type users and patient subjects_ - patient information forms are created or updated with the demographics data returned after patient identification.

### **NOT** included in this PR:
* [CARDS-1339](https://phenotips.atlassian.net/browse/CARDS-1339): _Patient users should only be able to access the PROMs ui_ (subtask of [CARDS-1344](https://phenotips.atlassian.net/browse/CARDS-1344)) which is dependent on [CARDS-1338](https://phenotips.atlassian.net/browse/CARDS-1338): _Create the patient user type_ - not yet implemented.
* Patient-type user authentication - [CARDS-1337](https://phenotips.atlassian.net/browse/CARDS-1337)
* Final questionnaire structure ([CARDS-1355](https://phenotips.atlassian.net/browse/CARDS-1355)) and access rights on specific questions / sections - ([CARDS-1346](https://phenotips.atlassian.net/browse/CARDS-1346))

### Testing instructions
* Start cards in the `proms` run mode
* Navigate to `/Proms.html/Cardio`
* Log in
* Follow the mock patient identification screen instructions; the following test "patients" are currently recognized by the mock authenticator:
```
[
  {
    last_name: "Addison",
    first_name: "John",
    date_of_birth: "1970-01-01",
    mrn: "1234567",
    health_card: "1234567890AB",
    visit_number: "11"
  }, {
    last_name: "Bennet",
    first_name: "Mary",
    date_of_birth: "1971-02-27",
    mrn: "2345678",
    health_card: "2345678901CD",
    visit_number: "22"
  }, {
    last_name: "Coleman",
    first_name: "Paul",
    date_of_birth: "1972-03-31",
    mrn: "3456789",
    health_card: "3456789012EF",
    visit_number: "33"
  }
];
```
* Also navigate to `/Proms.html` and `/Proms.html/NonExistingProm` to test what happens when the patient goes to the wrong Proms link
* Check that after patient identification there's one Patient Information form for that patient subject, that gets created / updated with the patient data
* Leave some mandatory questions unanswered and check that you're warned about them when you reach the end
* Interrupt answering the questionnaires, navigate away from the page, then return and check that you can pick up where you left off after patient re-identification
* **Note 1: The patient identification functionality is only a mock-up included here to support the survey workflow. There is currently no patient-type user and the patient is not truly authenticated.** Patient authentication will be covered separately by CARDS-1337 and CARDS-1338.
* **Note 2: The questionnaires are not in their final form**, and they will be improved as part of CARDS-1355 (Reviiew PROMs questionnaire) and CARDS-1346 (Configure permissions to prevent patient users from accessing the computed interpretation of their answers). **Testing should focus on the workflow, not on the questionnaire content.** Issues with the questionnaires should be reported as comments on the jira task CARDS-1355.

### Rreviewing instructions
* Reviewing commit by commit rather than file by file is recommended
* The `MockPatientIdentification` is a quick mock-up  included here to support the survey workflow. There is currently no patient-type user and the patient is not truly authenticated. Patient authentication will be covered separately by CARDS-1337 and CARDS-1338, and MockPatientIdentification` will be renamed and modified to support it at that time.

sdumitriu committed:
CARDS-1344 / CARDS-1363: Data model for proms

sdumitriu committed:
CARDS-1344 / CARDS-1339: Patient users should only be able to access the PROMs ui
Added a homepage node; no UI and no access rights yet

sdumitriu committed:
CARDS-1344 / CARDS-1362: Frontend module for the proms runmode

sdumitriu committed:
CARDS-1344 / CARDS-1364: Default QuestionnaireSet resource

marta- committed:
CARDS-1344 / CARDS-1363: Data model for proms & CARDS-1364: Default QuestionnaireSet resource
Added an `order` number to each step of a questionnaireSet

marta- committed:
CARDS-1340: The PROM UI should guide patient users through the list of questionnaires they must fill out

marta- committed:
CARDS-1358: Create a mock patient authentication component
Incomplete - login screen only

marta- committed:
CARDS-1365: The PROM UI should display the First and last name of the identified patient user at the top of every page
First attempt. Currently relies on the mock patient identification and may need to be updated after the actual patient identification is implemented.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,292 @@
+  const [ questionnaires, setQuestionnaires ] = useState();
+
+  // Step -1: haven't started yet, welcome screen
+  // Step i = 0 ... # of questionnaires-1 : filling out questionnaire 1
```
```suggestion
  // Step i = 0 ... # of questionnaires-1 : filling out questionnaire #i
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/MockPatientIdentification.jsx:
```diff
@@ -0,0 +1,154 @@
+              <Grid item>
+                <FormControl margin="normal" fullWidth>
+                  <InputLabel htmlFor="j_mrn">MRN</InputLabel>
+                  <Input id="j_mrn" name="j_mrn" autoComplete="off" type="number" onChange={event => setMrn(event.target.value)}/>
```
With `type=number` it looks weird with the up/down arrows that can increase/decrease the value. Can it be a simple input with number enforcing javascript instead?

marta- replied:
Yes we can. This is not something to worry about at this time, please see testing and review instructions.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/MockPatientIdentification.jsx:
```diff
@@ -0,0 +1,154 @@
+         </Grid>
+         <Grid item xs={12}>
+            <FormControl margin="normal" required fullWidth>
+              <InputLabel htmlFor="j_dob">Date of birth</InputLabel>
```
Looks weird if you focus in then out without typing something:

![image](https://user-images.githubusercontent.com/88663/130847219-9b2dea1a-e19b-4bea-a558-ccd2a124f5ec.png)

marta- replied:
This is not something to worry about at this time, please see testing instructions.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,292 @@
+  if (!id) {
+    return (
+      <QuestionnaireSetScreen className={classes.screen} items={[
+        <Typography variant="h4" color="textSecondary">You do not have any pending surverys</Typography>
```
I only get this screen, for all 3 test users. How do I get to actually fill in a questionnaire set?

sdumitriu replied:
Ah, I think the SamplePatient is needed. Let me try that.

sdumitriu replied:
OK, figured it out, I just had to read the instructions and open `/Proms.html/Cardio` instead of `/Proms`.

marta- committed:
CARDS-1358: Create a mock patient authentication component
Added Patient and Visit subject creation upon patient identification, if the subjects don't already exist
The implementation of this task is still incomplete as it is missing the authentication of an actual patient-type user

marta- committed:
CARDS-1349: Sync patient-type users and patient subjects

marta- committed:
CARDS-1358: Create a mock patient authentication component
Minor ui improvement of the form: fixed overlapping labels, added placeholder

marta- committed:
CARDS-1340: The PROM UI should guide patient users through the list of questionnaires they must fill out
When a questionnaire has already been filled out, skip it

marta- committed:
CARDS-1340: The PROM UI should guide patient users through the list of questionnaires they must fill out
When a questionnaire has already been partially filled out but is incomplete, allow editing it

marta- committed:
CARDS-1340: The PROM UI should guide patient users through the list of questionnaires they must fill out
On the last screen, if any questionnaires are incomplete, inform the user and enable them to go back and fill out the missing answers

marta- committed:
CARDS-1340: The PROM UI should guide patient users through the list of questionnaires they must fill out
If the user is back to the start because they were directed to add missing answers, dont't show the welcome screen and skip to the next step without them pressing start

marta- committed:
CARDS-1358: Create a mock patient authentication component
Minor refactoring

marta- committed:
CARDS-1344: As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires
* show the list of available proms if no promid is specified
* when no prom is found for the specified prom id, show a user-friendly message and a link to browse the list of available proms

marta- commented:
> > I followed the instructions and everything seems to be working fine. The only question I have is when I log into `/Proms.html` I get "You do not have any pending surveys". Is that expected? Would it be better if we offer them link(s) that is available at the moment e.g. Cardio link
> 
> Patients will get a link directly to the survey and they should follow that link, not easily pick a survey from the list that may not be relevant for them. I can make a very simple homepage with links, but I need to check with the client that this is what they want.
> 
> > Also, goin to `Proms.html/Lala gives me not so user friendly error
> > Can we change it to "There is no such survey, Please try" + what is available (e.g. link to Cardio).
> 
> Will rephrase the error message.

@sashaandjic :Addressed in 364f425 .

sdumitriu commented:
I'm not sure listing available sets is wanted or useful. As a patient going to the hospital, I should not be able to just start filling questionnaires that are not relevant to me. It is the doctor who decides what the patient should fill in.

marta- commented:
> I'm not sure listing available sets is wanted or useful. As a patient going to the hospital, I should not be able to just start filling questionnaires that are not relevant to me. It is the doctor who decides what the patient should fill in.

We'll present it as is and we can change it after if they object.

sashaandjic commented:
tested
![image](https://user-images.githubusercontent.com/23243244/131232678-7f89939f-f108-433c-b00a-0701bec2f35b.png)
![image](https://user-images.githubusercontent.com/23243244/131232694-7247a30d-755a-45d5-9b44-61610ada00c5.png)
![image](https://user-images.githubusercontent.com/23243244/131232729-de04acc8-f888-44f0-9c8e-e089a5156e9a.png)
![image](https://user-images.githubusercontent.com/23243244/131232751-992698a8-0172-497c-b67c-9d3bdc8cbf63.png)
![image](https://user-images.githubusercontent.com/23243244/131232764-cf53f554-5c1d-4233-b595-8b878e5fb683.png)
![image](https://user-images.githubusercontent.com/23243244/131232773-e091bf96-6c58-48ef-ac86-ffcecc4a7a23.png)

marta- commented:
> > > I'm not sure listing available sets is wanted or useful. As a patient going to the hospital, I should not be able to just start filling questionnaires that are not relevant to me. It is the doctor who decides what the patient should fill in.
> > 
> > 
> > We'll present it as is and we can change it after if they object.
> 
> I agree with Sergiu - there is no option for patients _deciding_ which surveys they want to complete. These are prescribed by the clinician who determines the appropriateness

Updated in commit f1f8b8e .

marta- committed:
CARDS-1344: As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires
Reverting some of the changes form the previous commit: the patient user should only have access to surveys via direct link, they should not be given the ability to browse existing surveys.

k-pace commented:
**Minor comment:** 

**Other to-dos/suggestions:** 
- [  ] hide scoring feedback at bottom of screen from patients
- [  ] increase font size in Smoking survey:
![image](https://user-images.githubusercontent.com/33154316/131384107-ddc6653e-4b1f-4eec-afdf-08b65fe2578a.png)

**Discussion:**
- Should we allow a user to progress to the next survey without completing all mandatory questions? IMHO I would expect to not let them proceed and scroll to the earliest incomplete answer upon clicking proceed. Thoughts?
- in the GAD and PHQ survey intro text, should 'Pearls' be 'Perils'?

marta- commented:
> **Minor comment:**
> 
> **Other to-dos/suggestions:**
> 
> * [  ] hide scoring feedback at bottom of screen from patients

This falls under access rights on specific questions / sections - ([CARDS-1346](https://phenotips.atlassian.net/browse/CARDS-1346)) - not included in this PR

> * [  ] increase font size in Smoking survey:
>   ![image](https://user-images.githubusercontent.com/33154316/131384107-ddc6653e-4b1f-4eec-afdf-08b65fe2578a.png)

This falls under reviewing PROMS questionnaires ([CARDS-1355](https://phenotips.atlassian.net/browse/CARDS-1346)); suggestion added as a comment to 1355 on Jira.

> 
> **Discussion:**
> 
> * Should we allow a user to progress to the next survey without completing all mandatory questions? IMHO I would expect to not let them proceed and scroll to the earliest incomplete answer upon clicking proceed. Thoughts?

I created https://phenotips.atlassian.net/browse/CARDS-1375, details need to be ironed out.

> * in the GAD and PHQ survey intro text, should 'Pearls' be 'Perils'?

I honestly don't know what "Pearls" is doing there, but that's what the original looks like: https://www.mdcalc.com/gad-7-general-anxiety-disorder-7#pearls-pitfalls. However, that part will not be displayed to the user and it will not be exported to the EPR (hiding it falls under  access rights on specific questions / sections - ([CARDS-1346](https://phenotips.atlassian.net/browse/CARDS-1346)) - not included in this PR.

sashaandjic reviewed (approved):
RE-TESTED

sashaandjic commented:
We need to clarify the "pearls" from the originals. It was probably a spelling error.
![image](https://user-images.githubusercontent.com/23243244/131909415-15afc90e-37ae-4094-8c9f-8c1c658d9279.png)

sashaandjic commented:
I am not sure if this was done on purpose, but when I click on these 2 links, the BACK button brings me back to the login page, rather than back to the survey page where links were initiated:
![image](https://user-images.githubusercontent.com/23243244/131909882-dee9e14b-42ea-40ba-8196-7652b31aa98b.png)

marta- commented:
> We need to clarify the "pearls" from the originals. It was probably a spelling error.

Patients will not see this information, it will be hidden as part of CARDS-1346. Questionnaire structure is out of scope for this PR.

marta- commented:
> I am not sure if this was done on purpose, but when I click on these 2 links, the BACK button brings me back to the login page, rather than back to the survey page where links were initiated:

Patients will not see this link. The whole section will be hidden as part of CARDS-1346, so in production this will be a non-issue.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSets.jsx:
```diff
@@ -0,0 +1,80 @@
+//
```
This seems unused, can it be deleted?

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+    crtFormId && nextStep();
+  }, [crtFormId]);
+
+  // At the last step, determine if the questionnaires have been ceompleted
```
```suggestion
  // At the last step, determine if the questionnaires have been completed
```

sdumitriu replied:
Or a bit more correct:

```suggestion
  // At the last step, reload the form data to determine if all the questionnaires have been completed
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+  // When something goes wrong:
+  const [ error, setError ] = useState("");
+  // Screen layout props
+  const [ screenType, setScreenType ] = useState ();
```
```suggestion
  const [ screenType, setScreenType ] = useState();
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+  useEffect(() => {
+    if (!subjectData || !questionnaires) return;
+    setComplete(Object.keys(subjectData || {}).filter(q => isFormComplete(q)).length == questionnaireIds.length);
+  }, [subjectData, questionnaires]);
```
It's not clear to me why the conditions are on `questionnaires` but the code itself uses `questionnaireIds`. Shouldn't `questionaireIds` be used throughout?

```suggestion
    if (!subjectData || !questionnaireIds) return;
    setComplete(Object.keys(subjectData || {}).filter(q => isFormComplete(q)).length == questionnaireIds.length);
  }, [subjectData, questionnaireIds]);
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+      .then((json) => {
+        let data = {};
+        questionnaireIds.forEach(q => {
+          if (json[questionnaires[q]?.title]?.[0]?.['jcr:primaryType'] == "cards:Form") {
```
Hopefully no visit will require filling the same questionnaire more than once.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+
+  let launchNextForm = () => {;
+    if (subjectData[nextQuestionnaire['@name']]) {
+      // Form already exists and is incomplete: prepare  to edit it
```
```suggestion
      // Form already exists and is incomplete: prepare to edit it
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+}
+
+function QuestionnaireSetScreen (props) {
+  let { items, ...rest } = props;
```
Why aren't you using `props.children` instead of the custom `items`?

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/MockPatientIdentification.jsx:
```diff
@@ -0,0 +1,332 @@
+  // if the subject doesn't exist, create it
+  const getSubject = (subjectType, pathPrefix, subjectId, subjectIdLabel, successCallback) => {
+    // If the patient doesn't yet have an MRN, or the visit doesn't yet have a number, abort mission
+    // Todo: after we find out if the MRN is not always assigned in the DHP,
```
```suggestion
    // TODO: after we find out if the MRN is not always assigned in the DHP,
```

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/MockPatientIdentification.jsx:
```diff
@@ -0,0 +1,332 @@
+  }
+
+  // Create a new subject if it's the first time we receive this identifier
+  const createSubject = (type, path, id, successCallback) => {
```
I don't like that patients will have the right to create subjects. It will have to do for the moment, but I'd like to look for a better solution once all the pieces fall into place. Right now I'm envisioning a special endpoint that gets or creates the correct patient/visit subjects, and only those, without any other access to subjects.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/QuestionnaireSet.jsx:
```diff
@@ -0,0 +1,407 @@
+  let createNextForm = (formId, successCallback) => {
+    if (!formId) return;
+    // Make a POST request to create a new form, with a randomly generated UUID
+    const URL = "/Forms/" + formId;
```
Just like for subjects, I would like in the end a more contained form creation endpoint.

sdumitriu commented on proms-resources/frontend/src/main/frontend/src/MockPatientIdentification.jsx:
```diff
@@ -0,0 +1,332 @@
+    patient && piDefinition && syncPatientInfo();
+  }, [patient, piDefinition]);
+
+  const syncPatientInfo = () => {
```
This will be done in the backend when the real authentication is implemented. Patients will have no write access to the PI form.

marta- committed:
Apply suggestions from code review

Co-authored-by: Sergiu Dumitriu <sdumitriu@users.noreply.github.com>

marta- commented:
Closing for now. Will be re-implemented properly when we have the patient authentication and backend synchronization in place.

sdumitriu commented:
I don't think this should have been closed. Other than the minor code changes, my three comments about permissions and data sync workflow were not change requests, but ideas for the future.

sdumitriu committed:
CARDS-1344: As a patient-type user, I can access a specific url to start filling out a sequence of questionnaires
Replace custom `items` prop with the React standard `children`
